File: sparkJava/src/main/java/com/bearhouse/AppendTimestamp.java
package com.bearhouse;

import com.sun.org.apache.xerces.internal.impl.xpath.regex.RegularExpression;

import org.apache.spark.SparkContext;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.rdd.RDD;
import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormat;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import scala.util.matching.Regex;

/**
 * Created by md186047 on 18/9/17.
 */
public class AppendTimestamp {

    private static final Logger log = LoggerFactory.getLogger(AppendTimestamp.class);

    public static void main(String args[]){

        log.info("Running Spark AppendTimestamp ...");

        if(args.length < 8 || args.length > 9){
            System.out.println("No right number of arguments. Please provide 8 arguments");
            System.exit(1);
        }


            String categoryName = args[0];
            String feedName = args[1];
            String partitionTs = args[2];
            String sourceRoot = args[3];
            String ingestRoot = args[4];
            String fileName = args[5];
            String sourceTimestampPatten = args[6];
            String dateFormat = args[7];
            final String delimiter = (args[8] == null ) ? "," : args[8];

            final SimpleDateFormat outputDateTimesFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
            String partitionPath = categoryName + "/" + feedName +"/" + partitionTs + "/";

            // Initialize Spark Content
            JavaSparkContext sc = new JavaSparkContext();


            final String outputTimestamp = parseTimestamp(fileName, sourceTimestampPatten, dateFormat);

            if(!(outputTimestamp == null )){

                // add timestamp to data
                String path = sourceRoot +  partitionPath;
                JavaRDD sourceDataRDD = sc.textFile(path);
                JavaRDD<String> ingestDataRDD = addSourceTs(sourceDataRDD, delimiter, outputTimestamp)
;
                // save the output
                ingestDataRDD.saveAsObjectFile(path);
            } else{
                System.exit(1);
            }

    }

    public static String parseTimestamp(String fileName, String tsRegex, String tsFormat) {
        Pattern regex = Pattern.compile(tsRegex);
        Matcher regexMatcher = regex.matcher(fileName);

        if (regexMatcher.find()) {
            String inputTsString = regexMatcher.group();

            SimpleDateFormat inputTsFormatter = new SimpleDateFormat(tsFormat);
            try {
                Date inputTs = inputTsFormatter.parse(inputTsString);

                SimpleDateFormat outputTsFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
                String outputFormattedTs = outputTsFormat.format(inputTs);
                return outputFormattedTs;

            } catch (ParseException e) {
                log.error(String.format("ERROR - Cannot convert timestamp {} based on format {}", inputTsString, tsFormat));
                e.printStackTrace();
            }
        }

        return null;
    }

    public static JavaRDD<String> addSourceTs(JavaRDD sourceDataRDD, final String delimiter, final String outputTimestamp){
        return sourceDataRDD.map(
            new Function<String, String>() {
                public String call(String line) throws Exception {
                    return line + delimiter + outputTimestamp;
                }

            }
        );
    }



}

File: sparkScala/src/main/scala/com/bearhouse/AppendTimestamp.scala
package com.bearhouse

import java.text.SimpleDateFormat
import java.util.logging.SimpleFormatter

import org.apache.log4j.LogManager
import org.apache.spark.SparkContext

import scala.util.matching.Regex

/**
  * Created by md186047 on 20/9/17.
  */
object AppendTimestamp {




  def main(args: Array[String]): Unit ={

    val log = LogManager.getRootLogger

    log.info("Running Spark AppendTimestamp with the following command line args (comma separated)")

    if(args.length < 8 && args.length > 9){
      System.out.println("Please provide 8 parameters");
    }

    try{
      val categoryName = args(0)
      val feedName = args(1)
      val partitionTs = args(2)
      val sourceRoot = args(3)
      val ingestRoot = args(4)
      val filename = args(5)
      val sourceTimestampPattern = args(6).r // .r for regular expression
      val sourceDateFormat = args(7)
      val delimiter = if (args.length ==8 )  "," else args(8)

      val outputDateTimeFormatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")

      // TOGET : how to combine a string
      val partitionPath = s"$categoryName/$feedName/$partitionTs"

      // init Spark Context
      val sc = new SparkContext()

      // timestamp matching
      val outputTimestamp = parseTimestamp(filename, sourceTimestampPattern, sourceDateFormat, outputDateTimeFormatter)

      if(outputTimestamp.nonEmpty) {

        // Add timestamp to data
        val sourceDataRDD = sc.textFile(sourceRoot + partitionPath)
        val ingestDateRDD = sourceDataRDD.map(line => appendTs(line, delimiter, outputTimestamp))

        // save output to a file
        ingestDateRDD.saveAsTextFile(ingestRoot + partitionPath)
      }
      // Fail if not match
      else System.exit(1)
    }
    catch {
      case e: Exception => {
        System.out.println(e)
        System.exit(1)
      }
    }
  }

  // TOGET: a function returns Option[dataType] rather than a dataType straightaway in order to check NULL value
  def parseTimestamp(filename: String, tsRegex: Regex, inputDateFormat: String, outputDateTimeFormatter: SimpleDateFormat): Option[String] = {

    // to get  timestamp string from file name
    val inputTsOption = tsRegex.findFirstMatchIn(filename)

    if(inputTsOption.nonEmpty) {
      val inputTsFormatter = new SimpleDateFormat(inputDateFormat)
      val inputTs = inputTsFormatter.parse(inputTsOption.get.toString())

      val outputFormattedTimestamp = outputDateTimeFormatter.format(inputTs)

      // TOGET: no return statement, just put an expression
      Some(outputFormattedTimestamp)
    }
    else None

  }


  def appendTs(line: String, delimiter: String, outputTimestamp: Option[String]): String = {
    line + delimiter + outputTimestamp
  }

}
